src/Tokenizer/Transformer/AmpersandTransformer.php

```diff
@@ -0,0 +1,58 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of PHP CS Fixer.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *     Dariusz Rumiński <dariusz.ruminski@gmail.com>
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace PhpCsFixer\Tokenizer\Transformer;
+
+use PhpCsFixer\Tokenizer\AbstractTransformer;
+use PhpCsFixer\Tokenizer\CT;
+use PhpCsFixer\Tokenizer\Token;
+use PhpCsFixer\Tokenizer\Tokens;
+
+/**
+ * Transforms ampersand T_AMPERSAND_FOLLOWED_BY_VAR_OR_VARARG and T_AMPERSAND_NOT_FOLLOWED_BY_VAR_OR_VARARG into single CT::T_AMPERSAND.
+ *
+ * @internal
+ */
+final class AmpersandTransformer extends AbstractTransformer
```

Comment:
we should think about adding Transformer as BC breaker. 
the end-users, having the custom fixers, will be affected by this, as each fixer would need to adopt to the changes introduced by new Transformer

Reply:
This transformer is only for PHP 8.1 which we do not support yet.

In other words - it's not changing any behaviour in any supported PHP version.

Reply:
- on PHP 7.4, PHP Tokenizer will consider `&` as single token of `'&'`
- on PHP 8.1, PHP Tokenizer will consider `&` as one of `[T_AMPERSAND_FOLLOWED_BY_VAR_OR_VARARG, '&']`, `[T_AMPERSAND_NOT_FOLLOWED_BY_VAR_OR_VARARG, '&']` or `'&'`

Now, the question - is it only that tokenizer will be more specific on existing usages of `&`, or is it that usage of `T_AMPERSAND_ ...` are NOT allowed until PHP 8.1?
- if first, Transformer would be a BC breaker.
- if later, we can introduce Transformer

Reply:
how many different tokens will `&` be represented by under native PHP 8.1 tokenizer? 2 tokens and one string representation, or only 2 tokens?
in previous message, I assumed that under 8.1 we have 3 forms.

If we gonna have only 2 forms, then it's the following:
- on 7.4 we have `'&'`, on 8.1 we have `T_AMPERSAND_FOLLOWED_BY_VAR_OR_VARARG` or `T_AMPERSAND_NOT_FOLLOWED_BY_VAR_OR_VARARG`, which one token is exactly what `'&'` was about earlier, and other stands for new feature.
For that, I would treat `T_AMPERSAND_FOLLOWED_BY_VAR_OR_VARARG` internally as alias for `'&'` and DO NOT mark it with same CustomToken as `T_AMPERSAND_NOT_FOLLOWED_BY_VAR_OR_VARARG`, as `T_AMPERSAND_NOT_FOLLOWED_BY_VAR_OR_VARARG` is used in different context and should be handled separately.

Treating `T_AMPERSAND_FOLLOWED_BY_VAR_OR_VARARG` internally as alias for `'&'` will help creators of custom fixers to not have a need to rewrite all their fixers from using `->equals('&')` to `->is(T_AMPERSAND_FOLLOWED_BY_VAR_OR_VARARG)`

On the long run, we should deprecate direct usage of `'&'` and rely always on 2 native `T_AMPERSAND_...` tokens

Reply:
> Now, the question - is it only that tokenizer will be more specific on existing usages of &, or is it that usage of T_AMPERSAND_ ... are NOT allowed until PHP 8.1?

Oh, I didn't make that clear, tokens `T_AMPERSAND_FOLLOWED_BY_VAR_OR_VARARG` and `T_AMPERSAND_NOT_FOLLOWED_BY_VAR_OR_VARARG` are introduced in PHP 8.1.

> how many different tokens will & be represented by under native PHP 8.1 tokenizer? 2 tokens and one string representation, or only 2 tokens?

From what I've checked (`&` as union type, function reference, param reference, bitwise operator) it will be always one of the 2 new kinds.

> For that, I would treat T_AMPERSAND_FOLLOWED_BY_VAR_OR_VARARG internally as alias for '&' and DO NOT mark it with same CustomToken as T_AMPERSAND_NOT_FOLLOWED_BY_VAR_OR_VARARG, as T_AMPERSAND_NOT_FOLLOWED_BY_VAR_OR_VARARG is used in different context and should be handled separately.

That was my first thought too, but then you cannot arbitrarily decide when adding new token that it will be `T_AMPERSAND_FOLLOWED_BY_VAR_OR_VARARG` as it can be the other one (we check in tests if expected code is parsed to the same collection we have after fixing) - see the hack added in `OperatorLinebreakFixer` in https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/5776/files.

Especially that it kinda seems broken: https://3v4l.org/h7C4Q/rfc#vgit.master_jit and https://3v4l.org/h7C4Q/rfc#vgit.master_jit - thus the idea of custom token.

Reply:
> hack added in `OperatorLinebreakFixer`

Not sure if hack, to be honest. 
The fact that we have 2 ifs - we have 2 new tokens, both doing completely different things and by accident using same '&' text representation. we can have 2 separated handling of it
The fact that you override `T_AMPERSAND_NOT_FOLLOWED_BY_VAR_OR_VARARG` token with same (?) token? not clear from code why. maybe add some in-code explanation, nobody will have time to guess why you do that.

> kinda seems broken

no clue what "kinda seems broken" means. please, be explicit, we talk about low-level internals of PHP, using unclear description is not helping to discuss.
also, the 2 links you provided are the same, i guess you wanted to show different ones.

Reply:
> The fact that you override `T_AMPERSAND_NOT_FOLLOWED_BY_VAR_OR_VARARG` token with same (?) token? not clear from code why. maybe add some in-code explanation, nobody will have time to guess why you do that.

The override is not with same token T_AMPERSAND_NOT_FOLLOWED_BY_VAR_OR_VARARG with T_AMPERSAND_FOLLOWED_BY_VAR_OR_VARARG or the other way.

"kinda seems broken" -> https://bugs.php.net/bug.php?id=81342 - there is a better example, than the 3v4l links anyway.